layers/tcp.go

```diff
@@ -299,9 +307,21 @@ func decodeTCP(data []byte, p gopacket.PacketBuilder) error {
 	if err != nil {
 		return err
 	}
-	return p.NextDecoder(gopacket.LayerTypePayload)
+	if p.DecodeApplicationLayers() {
+		return p.NextDecoder(tcp.NextLayerType())
```

Comment:
Does it make sense to attempt decoding of a TCP payload?  I'd assume that with its streaming nature, you have no guarantees of packet boundaries in TCP, meaning this could often fail.

I'm okay with doing this in UDP, since it's datagram-based, but for TCP I think it's still best to just use Payload, then rely on reassembly elsewhere for application-level stuff.  Otherwise, callers may expect magic from this library that it's not prepared to give.

Also, DecodeApplicationLayers() is somewhat strange of a name, since UDP application layers (at least DNS and one or two others) are decoded regardless.